SI-8092 More verify for f-interpolator

som-snytt opened the pull request:
Attempt to verify the nooks and crannies of the format string.

Allows all syntax in the javadoc, including arg indexes. If the
specifier after an arg has an index that doesn't refer to the arg,
a warning is issued and the missing `%s` is prepended (just as
for a part with a leading `%n`).

Other enhancements include detecting that a `Formattable` wasn't
supplied to `%#s`.

Error messages attempt to be pithy but descriptive.

som-snytt committed:
SI-8092 More verify for f-interpolator

Attempt to verify the nooks and crannies of the format string.

Allows all syntax in the javadoc, including arg indexes. If the
specifier after an arg has an index that doesn't refer to the arg,
a warning is issued and the missing `%s` is prepended (just as
for a part with a leading `%n`).

Other enhancements include detecting that a `Formattable` wasn't
supplied to `%#s`.

Error messages attempt to be pithy but descriptive.

adriaanm commented:
Review by @densh please.

densh commented on src/compiler/scala/tools/reflect/MacroImplementations.scala:
```diff
@@ -1,171 +1,296 @@
+          List(ids: _* )
+        )
+      Block(evals.toList, atPos(origApplyPos.focus)(expr)) setPos origApplyPos.makeTransparent
+      //q"..evals, ${Literal(Constant(fstring))}.format(..$ids)"
```
`q"..$evals; $fstring.format(..$ids)"`

densh commented on src/compiler/scala/tools/reflect/MacroImplementations.scala:
```diff
@@ -1,171 +1,296 @@
-    }
+    // require variants.nonEmpty
+    def checkType(arg: Tree, variants: Type*): Option[Type] =
+      checkType1(arg, variants: _*) orElse Some(variants(0))
```
checkType functions could use a better name. Maybe:

```
checkType0 --> findSubtype
checkType1 --> findConforming
checkType  --> findConformingOrFirst
```

densh commented on src/compiler/scala/tools/reflect/MacroImplementations.scala:
```diff
@@ -1,171 +1,296 @@
+   *
+   *  [index$][flags][width][.precision]conversion
+   */
+  def macro_StringInterpolation_f(parts: List[Tree], args: List[Tree], origApplyPos: c.universe.Position): Tree = {
```
Is it just me or this code has become a bit too nested?
I'd personally love to see something like:
1. rename `MacroImplementations` into `FInterpolator` (there are no more implementations in there)
2. make `def expandFInterpolator` an entry point without arguments
3. Move parsing of the macro prefix from `FastTrack` to top-level of the `FInterpolator`

This way you can flatten most of current `def macro_StringInterpolation_f` into the `FInterpolator` class and `expandFInterpolator` will be a clean entry point to that with minimal high-level code.  This scheme is very similar to what I have in quasiquotes and I think it's a win-win.

densh replied:
/cc @xeno-by

som-snytt replied:
I agree.  My focus was entirely on getting close to correctness.

I've only just discovered bundles. An amazing convenience.

densh commented on src/compiler/scala/tools/reflect/MacroImplementations.scala:
```diff
@@ -1,171 +1,296 @@
+        }
+        def cv(cc: Char) = cc match {
+          case 'b' | 'B' | 'h' | 'H' | 's' | 'S' =>
+            new Conversion(m, p) with GeneralXn
```
Why not `class FoobarXn(m, p) extends Conversion(m, p)` and `new FoobarXn(m, p)`? Current representation will create twice as much classes in the bytecode.

som-snytt commented on src/compiler/scala/tools/reflect/MacroImplementations.scala:
```diff
@@ -1,171 +1,296 @@
-      // 4) "...${smth}%n" => okay, equivalent to "...${smth}%s%n"
-      // 5) "...${smth}%%" => okay, equivalent to "...${smth}%s%%"
-      // 6) "...${smth}[%legalJavaConversion]" => okay, according to http://docs.oracle.com/javase/1.5.0/docs/api/java/util/Formatter.html
-      // 7) "...${smth}[%illegalJavaConversion]" => error
```
Maybe restore the comment in some form.

som-snytt committed:
SI-8092 Refactor f-interp

A denshish refactor makes the FormatInterpolator a nice bundle
that destructures its input and flattens out the classes to
give the code some elbow room. Everything shifts left.

The `checkType` method is refolded and renamed `pickAcceptable`.

An additional test case captures the leading edge test, that
a % should follow a hole, and which is the most basic
requirement.

som-snytt commented:
Took all of @densh 's good advice.  Many thanks for taking the time.  It's non-trivial LOC.

I was also looking at quasiquotes and parser, it's all pretty neat reading. So I esp appreciate time away from that for this.

densh commented on src/compiler/scala/tools/reflect/FormatInterpolator.scala:
```diff
@@ -0,0 +1,313 @@
+  private def fail(msg: String) = c.abort(c.enclosingPosition, msg)
+
+  def interpolate: Tree = c.macroApplication match {
+    case q"$_(..$parts).f(..$args)" =>
```
Unfortunately we can't use quasi-quotes in the 2.11 compiler code. Use treeInfo.Applied instead.

densh commented on src/compiler/scala/tools/reflect/FormatInterpolator.scala:
```diff
@@ -0,0 +1,313 @@
+      }
+      if (badlyInvoked) c.macroApplication else interpolated(parts, args)
+    case other =>
+      fail(s"Unexpected application ${showRaw(other)}")
```
global.abort is probably more appropriate in situations like that.

som-snytt committed:
SI-8092 Review cleanup, no qq

No crazyquoting. Use global.abort on total fail.

Remove unnecessary usage of varargs Apply, per review.

som-snytt commented:
Push without crazyquotes. They're totally crazy!

som-snytt commented:
Pushed squashed without the (legacy) varargs args for `Apply`.  It took me a while to chase down what `Apply` is and which I want to invoke.

In the "refactor" commit, I changed `contexts.Context` to `runtime.Context` to look like the qq code.  I would have done a bit more to homogenize if I had time.

densh commented:
LGTM
